source/benchmarking/odm-benchmarking.md

```diff
@@ -0,0 +1,393 @@
+
+## Benchmark execution phases and measurement
+
+All benchmark tasks will be conducted via a number of iterations. Each iteration will be timed and will generally
+include a large number of individual ODM operations.
+
+The measurement is broken up this way to better isolate the benchmark from external volatility. If we consider the
+problem of benchmarking an operation over many iterations, such as 100,000 model insertions, we want to avoid two
+extreme forms of measurement:
+
+- measuring a single insertion 100,000 times -- in this case, the timing code is likely to be a greater proportion of
+    executed code, which could routinely evict the insertion code from CPU caches or mislead a JIT optimizer and throw
+    off results
+- measuring 100,000 insertions one time -- in this case, the longer the timer runs, the higher the likelihood that an
+    external event occurs that affects the time of the run
+
+Therefore, we choose a middle ground:
+
+- measuring the same 10,000 insertions over 10 iterations -- each timing run includes enough operations that insertion
+    code dominates timing code; unusual system events are likely to affect only a fraction of the 10 timing measurements
+
+With 10 timings of inserting the same 10,000 models, we build up a statistical distribution of the operation timing,
+allowing a more robust estimate of performance than a single measurement. (In practice, the number of iterations could
+exceed 10, but 10 is a reasonable minimum goal.)
+
+Because a timing distribution is bounded by zero on one side, taking the mean would allow large positive outlier
+measurements to skew the result substantially. Therefore, for the benchmark score, we use the median timing measurement,
+which is robust in the face of outliers.
+
+Each benchmark is structured into discrete setup/execute/teardown phases. Phases are as follows, with specific details
+given in a subsequent section:
+
+- setup -- (ONCE PER TASK) something to do once before any benchmarking, e.g. construct a model object, load test data,
+    insert data into a collection, etc.
+- before operation -- (ONCE PER ITERATION) something to do before every task iteration, e.g. drop a collection, or
+    reload test data (if the test run modifies it), etc.
+- do operation -- (ONCE PER ITERATION) smallest amount of code necessary to execute the task; e.g. insert 10,000 models
+    one by one into the database, or retrieve 10,000 models of test data from the database, etc.
+- after operation -- (ONCE PER ITERATION) something to do after every task iteration (if necessary)
+- teardown -- (ONCE PER TASK) something done once after all benchmarking is complete (if necessary); e.g. drop the test
+    database
+
+The wall-clock execution time of each "do operation" phase will be recorded. We use wall clock time to model user
+experience and as a lowest-common denominator across ODMs. Iteration timing should be done with a high-resolution
```

Comment:
If this is a spec doc, we may want to use consistent formatting for admonitions. For example, this might be "SHOULD". Elsewhere you MAY want to use "MUST" 😄